### PR TITLE
Initial JA3 integration

### PR DESCRIPTION
**Pull request by thus**
This PR contains an initial JA3 integration (https://github.com/salesforce/ja3). This means that there are some features that are not done yet, and that will (hopefully) be added in the future, but it contains the core functionality.

Features implemented:
* Decode more of the TLS CLIENT_HELLO packet, and generate a JA3 string and a JA3 hash (md5sum of the JA3 string).
* Detection (mpm) keyword 'ja3_hash' for matching on the JA3 hash.
* Detection (mpm) keyword 'ja3_str' for matching on the entire or parts of the JA3 string.
* JSON logger for JA3.
* JA3 metadata in alerts.
* Lua functions 'Ja3GetHash' and 'Ja3GetStr' to access the JA3 buffers from Lua scripts.
* 'ja3_hash' and 'ja3_str' as custom fields in the TLS JSON logger.

As I briefly discussed with @inliniac, I also want to implement 'ja3_desc'. To do this I want to read a JSON file with JA3 fingerprints (hash, string and description) when Suricata starts and on rule reload. This file should contain all the known fingerprints and enable us to match on part of the description, e.g. "ja3_desc; content:"Firefox";). It should also enable us to match on all unknown JA3 fingerprints. Since the de_ctx is not available when decoding the TLS packet, I have not added this feature yet, but it is something that (IMO) would add a lot of value to the JA3 integration.

As of now, I have choose to disable generating JA3 fingerprints by default. I have added an entry to enable it to the suricata.yaml file in the repo, but this means that it won't be enabled before the user updates their Suricata config. Please let me know if you disagree on doing this.

Redmine:
https://redmine.openinfosecfoundation.org/issues/2192

Prscript:
- PR thus-pcap: https://buildbot.openinfosecfoundation.org/builders/thus-pcap/builds/143
- PR thus: https://buildbot.openinfosecfoundation.org/builders/thus/builds/143

### CODE REVIEW

**Comment by thus** [comment]
Changes:
* Fixed another compiler warning..
* Rebased with master

File: doc/userguide/output/eve/eve-json-format.rst [hunk 1]
```diff
@@ -254,6 +254,8 @@ In addition to this, custom logging also allows the following fields:
 
 * "certificate": The TLS certificate base64 encoded
 * "chain": The entire TLS certificate chain base64 encoded
+* "ja3_hash": The JA3 hash, which is an md5sum of the JA3 string
```
Comment (norg) [inline doc/userguide/output/eve/eve-json-format.rst]: a m5sum instead of an md5sum?

File: doc/userguide/output/eve/eve-json-format.rst [hunk 2]
```diff
@@ -319,3 +321,21 @@ Example of TFTP logging:
       "file": "rfc1350.txt",
       "mode": "octet"
    }
+
+Event type: JA3
+---------------
+
+* "hash": The JA3 hash, which is an md5sum of the JA3 string
```
Comment (norg) [inline doc/userguide/output/eve/eve-json-format.rst]: a m5sum instead of an md5sum?

File: doc/userguide/output/eve/eve-json-output.rst [hunk 3]
```diff
@@ -107,13 +107,27 @@ YAML::
...
 
 The default is to log certificate subject and issuer. If ``extended`` is
 enabled, then the log gets more verbose.
 
 By using ``custom`` it is possible to select which TLS fields to log.
 
+JA3
+~~~
+
+JA3 fingerprints are logged one record per TLS session.
```
Comment (norg) [inline doc/userguide/output/eve/eve-json-output.rst]: on record? or did I misread this sentence?

**Comment by norg** [comment]
is there a specific reason why md5sum is used instead of newer sha2?

File: doc/userguide/output/eve/eve-json-format.rst (hunk 1 above)
Comment (thus) [reply doc/userguide/output/eve/eve-json-format.rst]: What is a 'm5sum'? Isn't 'md5sum' correct here?

File: doc/userguide/output/eve/eve-json-format.rst (hunk 2 above)
Comment (thus) [reply doc/userguide/output/eve/eve-json-format.rst]: Same as above :)

File: doc/userguide/output/eve/eve-json-output.rst (hunk 3 above)
Comment (thus) [reply doc/userguide/output/eve/eve-json-output.rst]: I was thinking it as one JSON record per TLS session. Should it be changed to something else?

**Comment by thus** [comment]
The reason why md5 was chosen, is because that's what the authors of JA3 has used for their implementation. I have no idea why they chose md5 over other (newer) hash algorithms, I just thought it was best to use the same since this PR implements support for JA3 :)

**Comment by victorjulien** [comment]
As discussed on chat, this has some issues that need to be addressed.

**Comment by thus** [comment]
Replaced by #3215

**Pull request closed without merging**